src/Week7/HW7.hs

```diff
@@ -0,0 +1,240 @@
+  |otherwise = partitionRd v >>= selectSubvector i
+
+selectSubvector :: Ord a => Int -> (Vector a, a, Vector a) -> Rnd (Maybe a)
+selectSubvector i (v1, p, v2)
+  |i < l  = select i v1
+  |i == l = pure (Just p)
+  |i > l  = select (i-l-1) v2
+  where
+    l = length v1
+
+
+-- Exercise 10 ----------------------------------------
+
+-- Implement allCards, which returns a Vector of Cards grouped by suit and arranged from Two to Ace
+allCards :: Deck
+allCards = [Card l s | l <- toEnum <$> helperVec, s <- V.singleton Spade]
+        <> [Card l s | l <- toEnum <$> helperVec, s <- V.singleton Heart]
+        <> [Card l s | l <- toEnum <$> helperVec, s <- V.singleton Club]
+        <> [Card l s | l <- toEnum <$> helperVec, s <- V.singleton Diamond]
+
+helperVec :: (Num a) => Vector a
+helperVec = V.iterateN 13 (+1) 0 -- vector [0,1,...,12]
+
+-- This should return a new Deck that contains all of the cards from allCards, but in a random order
+newDeck :: Rnd Deck
+newDeck =  shuffle allCards
+
+-- Exercise 11 ----------------------------------------
+
+-- Define a function that takes in a Deck and gives you the head and tail (or Nothing if the Deck is empty)
+-- The vector module already has such a function!
+nextCard :: Deck -> Maybe (Card, Deck)
+nextCard = V.uncons
+
+-- Or, more explicitly:
+nextCard' :: Deck -> Maybe (Card, Deck)
+nextCard' v
+  |l <= 0    = Nothing
+  |otherwise = Just (V.head v, V.tail v)
+  where l = length v
+
+-- Exercise 12 ----------------------------------------
+-- Implement a function that draws n cars from the given Deck
+
+getCards :: Int -> Deck -> Maybe ([Card], Deck)
+getCards i d
+  |i >= length d || i < 0  = Nothing
+  |otherwise  = Just $ helper i d
+    where helper 0 v = ([], v)
+          helper n v = (V.head v : fst rest, snd rest)
+            where rest = helper (n-1) (V.tail v)
+
+
+-- Exercise 13 ----------------------------------------
+-- Simulation of a betting game. This one was given to us!
+
+data State = State { money :: Int, deck :: Deck }
+
+repl :: State -> IO ()
+repl s@State{..} | money <= 0  = putStrLn "You ran out of money!"
+                 | V.null deck = deckEmpty
+                 | otherwise   = do
+  putStrLn $ "You have \ESC[32m$" ++ show money ++ "\ESC[0m"
+  putStrLn "Would you like to play (y/n)?"
+  cont <- getLine
+  if cont == "n"
+  then putStrLn $ "You left the casino with \ESC[32m$"
+           ++ show money ++ "\ESC[0m"
+  else play
+    where deckEmpty = putStrLn $ "The deck is empty. You got \ESC[32m$"
+                      ++ show money ++ "\ESC[0m"
+          play = do
+            putStrLn "How much do you want to bet?"
+            amt <- read <$> getLine
+            if amt < 1 || amt > money
+            then play
+            else do
+              case getCards 2 deck of
+                Just ([c1, c2], d) -> do
+                  putStrLn $ "You got:\n" ++ show c1
+                  putStrLn $ "I got:\n" ++ show c2
+                  case () of
```

Comment:
Theres [MultiwayIf](https://downloads.haskell.org/ghc/8.8.3/docs/html/users_guide/glasgow_exts.html#extension-MultiWayIf) if you don't mind using another language extension :smile: